### PR TITLE
Various fixes in QualifierPolymorphism

### PR DESCRIPTION
**Pull request by smillst**
Fixes the crash reported in #738, but the Issue738.java test case has another crash. (See #701)

### CODE REVIEW

File: framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java [hunk 1]
```diff
@@ -192,7 +194,11 @@ public void annotate(MethodInvocationTree tree, AnnotatedExecutableType type) {
 
         Map<AnnotationMirror, Set<? extends AnnotationMirror>> matchingMapping = collector.visit(arguments, requiredArgs);
 
-        if (type.getReceiverType() != null) {
+        // for super() and this() method calls, getReceiverType(tree) does not return the correct
+        // type. So, just skip those.  This is consistent with skipping receivers of constructors
+        // below.
+        if (type.getReceiverType() != null && !(TreeUtils.isSuperCall(tree))
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java]: Why the parens around isSuperCall?

File: framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java [hunk 2]
```diff
@@ -231,6 +237,13 @@ public void annotate(AnnotatedExecutableType functionalInterface, AnnotatedExecu
 
         List<AnnotatedTypeMirror> args = functionalInterface.getParameterTypes();
         List<AnnotatedTypeMirror> requiredArgs = memberReference.getParameterTypes();
+        if (args.size() == requiredArgs.size() + 1) {
+            // If the member reference is a reference to an instance method of an arbitrary
+            // object, then first parameter of the functional interface corresponds to the
+            // receiver of the member reference.
+            requiredArgs = new ArrayList<>(requiredArgs);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java]: Isn't this a rather inefficient way of doing this? This first copies requiredArgs into the new array, and then has to move everything over to add a first element.

File: framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java [hunk 3]
```diff
@@ -392,8 +417,7 @@ protected Void scan(AnnotatedTypeMirror type, Void p) {
 
             assert actualType.getKind() == type.getKind();
             type = (AnnotatedDeclaredType) AnnotatedTypes.asSuper(types, atypeFactory, type, actualType);
-            // TODO: type is null if type is an intersection type
-            // assert type != null;
+
             if (type == null) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java]: Now that intersection types are already explicitly handled at the beginning of this method, can this be strengthened to an errorAbort if the type is null?

**Comment by wmdietl** [comment]
Just a few small comments.

File: framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java (hunk 3 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/util/QualifierPolymorphism.java]: Not yet.  asSuper returns null erroneously in other cases.  In the new asSuper, it is an ErrorAbort if asSuper returns null and therefore this (and many other) null checks will be removed.  The new asSuper ErrorAborts is how I found this batch of bugs.

**Comment by smillst** [comment]
Merged via 47470dc

**Pull request closed without merging**